File: verifier/src/main/kotlin/tech/thessemaj/deviceintelligence/verifier/Hkdf.kt
package tech.thessemaj.deviceintelligence.verifier

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * HKDF-SHA256 (RFC 5869), Extract-then-Expand, via HmacSHA256. Byte-for-byte
 * equivalent to the native dicore/crypto/hkdf.cpp (empty salt -> 32 zero bytes;
 * T(i) = HMAC(PRK, T(i-1)||info||i)). Extracted from TokenCryptoV2 so it can be
 * KAT'd directly against the RFC vectors.
 */
internal object Hkdf {
    fun sha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, outLen: Int): ByteArray {
        require(outLen in 0..(255 * 32)) { "HKDF outLen out of range: $outLen" }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(if (salt.isEmpty()) ByteArray(32) else salt, "HmacSHA256"))
        val prk = mac.doFinal(ikm)                    // Extract

        mac.init(SecretKeySpec(prk, "HmacSHA256"))    // Expand
        val out = ByteArray(outLen)
        var t = ByteArray(0)
        var pos = 0
        var counter = 1
        while (pos < outLen) {
            mac.update(t)
            mac.update(info)
            mac.update(counter.toByte())
            t = mac.doFinal()                          // resets the Mac to its keyed state
            val take = minOf(t.size, outLen - pos)
            System.arraycopy(t, 0, out, pos, take)
            pos += take
            counter++
        }
        return out
    }
}

File: verifiers/python/src/deviceintelligence_verifier/hkdf.py
"""HKDF-SHA256 (RFC 5869), Extract-then-Expand. Empty salt -> 32 zero bytes."""
import hashlib, hmac

def sha256(ikm: bytes, salt: bytes, info: bytes, out_len: int) -> bytes:
    if not 0 <= out_len <= 255 * 32:
        raise ValueError(f"HKDF outLen out of range: {out_len}")
    prk = hmac.new(salt if salt else b"\x00" * 32, ikm, hashlib.sha256).digest()
    out, t, pos, counter = b"", b"", 0, 1
    while pos < out_len:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        take = min(len(t), out_len - pos)
        out += t[:take]
        pos += take
        counter += 1
    return out
